core/tests/mod.rs

```diff
@@ -0,0 +1,646 @@
+            }
+        }),
+    );
+    let sumeragis = sumeragis.lock().unwrap();
+    let sumeragi = sumeragis.get(&peer_ids[0]).unwrap();
+
+    for _ in 0..300 {
+        if sumeragi
+            .apply_wsv(WorldStateView::clone)
+            .block_hashes_after_hash(None)
+            .len()
+            == 1
+        {
+            if let Some(hash) = *pb_hash.lock().unwrap() {
+                assert_eq!(
+                    hash,
+                    sumeragi
+                        .apply_wsv(WorldStateView::clone)
+                        .block_hashes_after_hash(None)[0]
+                );
+                return;
+            }
+        }
+        sleep_ms(100);
+    }
+    panic!("Sumeragi did not submit genesis.");
+}
+
+#[test]
+fn sumeragi_all_peers_commit_genesis() {
+    init_telemetry();
+
+    /* Can't have 1 peer, then block is not sent. Can't have < 4 peers because non genesis peers get upset. */
+    let (keypairs, peer_ids, configs, _queues, sumeragis) =
+        set_up_network_with_genesis(4, false, Arc::new(move |_packet, _peer_id, _sumeragis| {}));
+    let sumeragis = sumeragis.lock().unwrap();
+    let sumeragi = sumeragis.get(&peer_ids[0]).unwrap();
+    let other_key = keypairs[1].clone();
+    let other_config = configs[1].clone();
+
+    let genesis_block: VersionedCommittedBlock = {
+        let kura = Kura::blank_kura_for_testing();
+        let wsv = WorldStateView::new(genesis_world(peer_ids.clone()), Arc::clone(&kura));
+
+        assert_eq!(wsv.height(), 0);
+        assert_eq!(wsv.latest_block_hash(), None);
+
+        let genesis_network = GenesisNetwork::test(other_config, true).unwrap();
+        let transactions = genesis_network.transactions;
+        assert!(
+            !transactions.is_empty(),
+            "Genesis transaction set contains no valid transactions"
+        );
+
+        let block = BlockBuilder {
+            transactions,
+            event_recommendations: Vec::new(),
+            view_change_index: 0,
+            committed_with_topology: network_topology::Topology::new(peer_ids.clone()),
+            key_pair: other_key,
+            wsv,
+        }
+        .build();
+
+        assert!(
+            block.rejected_transactions.is_empty(),
+            "Genesis transaction set contains invalid transactions"
+        );
+
+        block.commit_unchecked().into()
+    };
+
+    sumeragi.incoming_message(MessagePacket::new(
+        ProofChain::default(),
+        Message::BlockSyncUpdate(genesis_block.clone().into()),
+    ));
+
+    for _ in 0..300 {
+        if sumeragi
+            .apply_wsv(WorldStateView::clone)
+            .block_hashes_after_hash(None)
+            .len()
+            == 1
+        {
+            assert_eq!(
+                genesis_block.hash(),
+                sumeragi
+                    .apply_wsv(WorldStateView::clone)
+                    .block_hashes_after_hash(None)[0]
+            );
+            return;
+        }
+        sleep_ms(100);
```

Comment:
Why do we need to wait actually? Can we refactor this to some event listenning? Maybe in another issue